pages/api/retrieval.ts

```diff
@@ -0,0 +1,166 @@
+  }
+}
+
+/**
+ * @description
+ * This function will make sure that the incoming request is valid by enforcing a set of
+ * rules.
+ * Request parameters :
+ *  - The maxRecords exists and it ranges between 1 and 10.
+ *  - A formID exists and is valid i.e 78
+ * BearerToken :
+ *  - Must contains an email address
+ *  - It has not expired
+ *  - It exists in our Database
+ *  - The record found in the database includes the email, active, and template_id fields
+ *  - The email on the token record matches the email in the payload.
+ *  - The active flag boolean is true.
+ * @param handler - A method to retrieve form's responses
+ * @returns
+ */
+export const formResponsesReqValidator = (
+  handler: (res: NextApiResponse, formID: string, maxRecords: number) => void
+) => {
+  return async function (req: NextApiRequest, res: NextApiResponse): Promise<unknown> {
+    try {
+      //Default value to 10 if it's undefined
+      const { maxRecords = "10", formID } = req.query;
+      //Check that formID and maxRecords aren't repeated
+      if (Array.isArray(formID) || Array.isArray(maxRecords))
+        return res.status(400).json({ error: "Bad Request" });
+      //Get formID form the bearer token
+      if (!formID) return res.status(400).json({ error: "Bad Request" });
+      //Check an empty object or string
+      const expectedMaxRecords = parseInt(maxRecords);
+      //Range is 1- 10
+      if (expectedMaxRecords < 1 || expectedMaxRecords > 10) {
+        return res.status(400).json({ error: "Invalid paylaod value found maxRecords" });
+      }
+      //Get token from request
+      const token = getBearerToken(req);
+      //Verify the token
+      const bearerTokenPayload = jwt.verify(
+        token,
+        process.env.TOKEN_SECRET || ""
+      ) as BearerTokenPayload;
+      const { email } = bearerTokenPayload;
+      //Check if an active formUserRecord exists for the given bearerToken.
+      const { exists } = (await checkActiveFormUserRecord(formID, email as string)).rows[0];
+      if (exists) return handler(res, formID, expectedMaxRecords);
+      return res.status(403).json({ error: "Missing or invalid bearer token." });
+    } catch (err) {
+      //Token verification has failed
+      res.status(403).json({ error: "Missing or invalid bearer token or unknown error." });
+    }
+  };
+};
+/*@description
+ * It returns an active record from form_users table base upon a formID/template_id and an email.
+ * @param formID - The id of the form
+ * @param email - The email that is associated to the formID
+ * @returns A form_user's record with active field i.e [{ active: true }]
+ */
+const checkActiveFormUserRecord = async (formID: string, email: string): Promise<QueryResult> => {
+  //Retrieving a tokenRecord or return an empty array
+  return executeQuery(
+    await dbConnector(),
+    "SELECT EXISTS(SELECT 1 FROM form_users WHERE template_id = ($1) and email = ($2) and active = true)",
```

Comment:
Rather than doing a nested query like this . Just do the actual query and if there is a row return true otherwise false

Reply:
fixed